File: examples/iteration/kotlin/core/range.kt
package salvo.core.range

import salvo.*
import salvo.core.array.*
import salvo.core.bytes.*
import salvo.core.fs.*
import salvo.core.iterator.*
import salvo.core.list.*
import salvo.core.map.*
import salvo.core.set.*
import salvo.core.string.*

data class Range(
    val start: Int,
    val end: Int,
    val step: Int,
)

data class __Pass_Range(
    var step: Int,
    var end: Int,
    var i: Int,
)

fun iter__5(range: Range): __Pass_Range {
    return __Pass_Range(end = range.end, step = range.step, i = range.start)
}

fun next__10(__p: __Pass_Range): Union2<Int, Finished> {
    val next = __p.i
    return when {
        __p.step == 0 -> {
            U2_1<Finished, Int>(finished())
        }
        __p.step > 0 && __p.i >= __p.end -> {
            U2_1<Finished, Int>(finished())
        }
        __p.step < 0 && __p.i <= __p.end -> {
            U2_1<Finished, Int>(finished())
        }
        else -> {
            __p.i = __p.i + __p.step
            U2_2<Finished, Int>(emitted(next))
        }
    }.let { when (it) { is U2_1<*, *> -> U2_2<Int, Finished>(it.value as Finished); is U2_2<*, *> -> U2_1<Int, Finished>(it.value as Int); } }
}

fun range(start: Int, end: Int, step: Int): Range {
    return Range(start = start, end = end, step = step)
}

fun range__2(start: Int, end: Int): Range {
    val step = when {
        start < end -> {
            1
        }
        start > end -> {
            -1
        }
        else -> {
            0
        }
    }
    return range(start, end, step)
}

fun range__3(end: Int): Range {
    return range__2(0, end)
}

File: examples/actors/rust/core/range.rs
use crate::core_array::*;
use crate::core_bytes::*;
use crate::core_fs::*;
use crate::core_iterator::*;
use crate::core_list::*;
use crate::core_map::*;
use crate::core_set::*;
use crate::core_string::*;
use crate::unions::*;

#[derive(Clone, Debug, PartialEq)]
pub struct Range {
    pub start: i32,
    pub end: i32,
    pub step: i32,
}

#[derive(Clone, Debug, PartialEq)]
pub struct __Pass_Range {
    pub step: i32,
    pub end: i32,
    pub i: i32,
}

pub fn iter__5(range: &Range) -> __Pass_Range {
    return __Pass_Range { end: range.end, step: range.step, i: range.start };
}

pub fn next__7(__p: &mut __Pass_Range) -> Union2<i32, Finished> {
    let mut next = __p.i;
    return (match if __p.step == 0 {
        Union2::<Finished, i32>::U1(finished())
    } else if __p.step > 0 && __p.i >= __p.end {
        Union2::<Finished, i32>::U1(finished())
    } else if __p.step < 0 && __p.i <= __p.end {
        Union2::<Finished, i32>::U1(finished())
    } else {
        __p.i = __p.i + __p.step;
        Union2::<Finished, i32>::U2(emitted(next))
    } { Union2::U1(__v) => Union2::<i32, Finished>::U2(__v), Union2::U2(__v) => Union2::<i32, Finished>::U1(__v), });
}

pub fn range(start: i32, end: i32, step: i32) -> Range {
    return Range { start: start, end: end, step: step };
}

pub fn range__2(start: i32, end: i32) -> Range {
    let mut step = if start < end {
        1
    } else if start > end {
        -1
    } else {
        0
    };
    return range(start, end, step);
}

pub fn range__3(end: i32) -> Range {
    return range__2(0, end);
}
